Contract data

graydon opened the pull request:
This is a consolidated branch including changes for CAP-0047 and 0053 (covering CONFIG_SETTING and CONTRACT_DATA ledger entries, and all support code).

jonjove commented on src/transactions/CreateContractOpFrame.cpp:
```diff
@@ -57,7 +57,17 @@ CreateContractOpFrame::doApply(AbstractLedgerTxn& ltx)
+    else
+    {
+        // FIXME: need a different error code here.
+        innerResult().code(CREATE_CONTRACT_ALREADY_EXISTS);
```
Let's get this fixed before we merge.

graydon replied:
This (along with the op itself) has been removed now.

jonjove commented on src/rust/src/contract.rs:
```diff
@@ -2,31 +2,106 @@
+    ledger_entries: &Vec<XDRBuf>,
 ) -> Result<Vec<u8>, Box<dyn Error>> {
-    let arg_scvals = ScVec::read_xdr(&mut Cursor::new(args.as_slice()))?;
+    let contract_id = xdr::Hash::read_xdr(&mut Cursor::new(contract_id.data.as_slice()))?;
```
Nitpick: this function desperately needs whitespace, and probably some comments too.

graydon replied:
I'll do some factoring and commenting but I can't do much about the whitespace -- that's how rustfmt formats it!

leighmcculloch replied:
You should be able to add empty lines. I don't think rustfmt will remove them?

graydon replied:
Perhaps? Anyway I split it into several support functions, much clearer now.

jonjove reviewed (changes requested):
This is almost mergeable. I'd really like if we could clean up the history a bit, because it will make things easier to understand when we thoroughly review everything before we release.

leighmcculloch commented on src/protocol-next/xdr/Stellar-contract.x:
```diff
@@ -123,8 +123,8 @@ struct SCMapEntry
 
-typedef SCVal SCVec<>;
-typedef SCMapEntry SCMap<>;
+typedef SCVal SCVec<256000>;
```
We've gone in circles on this. I think I was the one who asked for this. Maybe I misread the room, but I thought when we discussed this in the protocol meeting some people agreed we should have reasonable bounds.

The reason I asked for reasonable bounds is it is how we communicate with XDR parsers/libs and other applications about what reasonable and safe XDR is. Applications can of course always specify some upper bound of data that they're willing to parse into memory, but that is arbitrary and provides no certainty or understanding to an application about how much memory they need to actually load the XDR.

I don't think we need bounds for everything, but if we can do it, everything that could exist inside a transaction should have a bound.

>What will happen if contracts try to exceed these bounds?

Also, it is worth noting that XDR provides bounds already, there is no such thing as an unbounded array in XDR. So your question "what will happen if contracts to to exceed these bounds?" exists no matter if we specify the bounds or not. All XDR arrays are bound to 4G elements. We're not setting a bound here, we're setting a more reasonable bound than is already present.

>This sounds like a dangerous foot-gun to developers.

Could you elaborate? There is nothing here that can't happen at some other bound, and there is probably a myriad of reasons a tx might fail or a contract might be rejected deployment.

jonjove replied:
Good point that all XDR arrays are bound to 4G elements. Given that, setting a lower bound might actually be better--contracts will hit the error condition before they have a huge amount of data.

The foot-gun I was referring to is the fact that a contract developer unaware of the limit might not hit it in any unit tests. So they are likely to handle the error incorrectly or not at all.

graydon replied:
I left these as 256000 for now. I don't expect this is the final state of the discussion but it's easy to change before we finalize.

graydon replied:
Ok, set to a constant.

graydon commented on src/invariant/BucketListIsConsistentWithDatabase.cpp:
```diff
@@ -90,12 +90,15 @@ struct EntryCounts
     {
         return mAccounts + mTrustLines + mOffers + mData + mClaimableBalance +
-               mLiquidityPool;
+               mLiquidityPool + mContractCode + mContractData;
```
Fixed.

graydon commented on src/invariant/ConservationOfLumens.cpp:
```diff
@@ -29,6 +29,14 @@ calculateDeltaBalance(LedgerEntry const* current, LedgerEntry const* previous)
         break;
     case DATA:
         break;
+#ifdef ENABLE_NEXT_PROTOCOL_VERSION_UNSAFE_FOR_PRODUCTION
```
No good reason, fixed.

graydon commented on src/transactions/RevokeSponsorshipOpFrame.cpp:
```diff
@@ -44,7 +45,7 @@ getAccountID(LedgerEntry const& le)
         return *le.ext.v1().sponsoringID;
     default:
-        abort();
+        throw std::runtime_error("invalid LedgerEntry type");
```
No, I think this was churn left over from previous iterations of this PR where there were account IDs on the new LEs. Reverted.

graydon commented on src/transactions/SponsorshipUtils.cpp:
```diff
@@ -197,7 +198,12 @@ computeMultiplier(LedgerEntry const& le)
+#ifdef ENABLE_NEXT_PROTOCOL_VERSION_UNSAFE_FOR_PRODUCTION
+    case CONTRACT_CODE:
+    case CONTRACT_DATA:
+    case CONFIG:
```
Ok, removed.

graydon commented on src/transactions/SponsorshipUtils.cpp:
```diff
@@ -220,7 +226,11 @@ isSubentry(LedgerEntry const& le)
+    case CONTRACT_CODE:
+    case CONTRACT_DATA:
         return true;
+#endif
```
Oops, added. Also returning `false` for both CONFIG and CONTRACT_DATA.

MonsieurNicolas replied:
which code path ends up calling this for Jump Cannon?

graydon replied:
I don't think anything but there's fairly high fan-in to it from many callers, and it seemed harmless / defensive to add (it is correct that it is not a subentry).

MonsieurNicolas replied:
I think it's actually more conservative to do what we do for `LIQUIDITY_POOL` here (and throw): we should not have any dependency on "SponsorshipUtils" from Jump Cannon.

graydon commented on src/bucket/BucketApplicator.cpp:
```diff
@@ -203,22 +203,31 @@ BucketApplicator::Counters::reset(VirtualClock::time_point now)
             mDataDelete + mClaimableBalanceUpsert + mClaimableBalanceDelete +
-            mLiquidityPoolUpsert + mLiquidityPoolDelete;
+            mLiquidityPoolUpsert + mLiquidityPoolDelete + mContractCodeUpsert +
+            mContractCodeDelete + mContractDataUpsert + mContractDataDelete;
```
Nice catch, fixed.

graydon commented on src/transactions/test/ContractTests.cpp:
```diff
@@ -2,6 +2,8 @@
 // under the Apache License, Version 2.0. See the COPYING file at the root
 // of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
 
+#include "crypto/SecretKey.h"
```
Oh, because the editor inserted them. It's "clever" that way, but sometimes too clever :)

graydon replied:
(fixed)

graydon commented on src/transactions/contracts/ContractUtils.cpp:
```diff
@@ -2,12 +2,13 @@
 // under the Apache License, Version 2.0. See the COPYING file at the root
 // of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
 
+#include "xdr/Stellar-transaction.h"
```
Same here. Fixed.

graydon commented on src/protocol-next/xdr/Stellar-ledger-entries.x:
```diff
@@ -589,6 +658,9 @@ enum EnvelopeType
+    ENVELOPE_TYPE_POOL_REVOKE_OP_ID = 7,
+    ENVELOPE_TYPE_TX_CREATE_CONTRACT_TX = 8,
+    ENVELOPE_TYPE_CONTRACT_ID = 9,
+    ENVELOPE_TYPE_CHILD_CONTRACT_ID = 10
```
This was removed along with the op for creating contracts.

graydon commented on src/protocol-next/xdr/Stellar-ledger-entries.x:
```diff
@@ -491,6 +493,59 @@ struct LiquidityPoolEntry
+
+enum ConfigSettingType
+{
+    CONFIG_TYPE_UINT32 = 1
```
Fixed both of these to count from 0.

jayz22 commented on src/rust/src/contract.rs:
```diff
@@ -2,31 +2,154 @@
+            LedgerEntryData::ContractData(ContractDataEntry {
+                key,
+                val,
+                contract_id,
```
If I understood correctly, here the ledger entry is destructured into three members of the struct. Shouldn't the order be `contract_id`, `key`, `val`?

graydon replied:
They're destructured by field name, not position.

graydon commented on src/transactions/SponsorshipUtils.cpp:
```diff
@@ -198,6 +199,10 @@ computeMultiplier(LedgerEntry const& le)
+#ifdef ENABLE_NEXT_PROTOCOL_VERSION_UNSAFE_FOR_PRODUCTION
+    case CONTRACT_DATA:
+        return 1;
+#endif
```
No, in consultation with @jonjove this is an unreachable code path for CONFIG.

graydon replied:
I asked @jonjove again about this today (sorry, there are a lot of cases to handle and this one keeps coming up because it's weird) and he suggested I can remove _both_ here and let them both fail because they should be rejected in `RevokeSponsorshipOpFrame::doCheckValid`. So I'll remove this.

MonsieurNicolas commented on src/ledger/LedgerTxnConfigSQL.cpp:
```diff
@@ -0,0 +1,362 @@
+// Copyright 2022 Stellar Development Foundation and contributors. Licensed
```
forgot our friend Visual Studio 😿

graydon replied:
fixed.

MonsieurNicolas commented on src/main/ApplicationImpl.cpp:
```diff
@@ -194,6 +195,16 @@ maybeRebuildLedger(Application& app, bool applyBuckets)
                 app.getLedgerTxnRoot().dropLiquidityPools();
                 break;
+#ifdef ENABLE_NEXT_PROTOCOL_VERSION_UNSAFE_FOR_PRODUCTION
+            case CONTRACT_DATA:
```
missing schema upgrade/schema version bump in Database.cpp

graydon replied:
fixed.

MonsieurNicolas commented on src/protocol-next/xdr/Stellar-transaction.x:
```diff
@@ -32,6 +32,13 @@ struct DecoratedSignature
 };
 
+// Ledger key sets touched by a smart contract transaction.
+struct Footprint
```
maybe use a qualified name like "LedgerFootprint" when I hear "footprint" I think "certificate footprint" (sorry, this should be a CAP comment)

graydon replied:
Ok, changed to `LedgerFootprint`

graydon commented on src/ledger/LedgerTxn.cpp:
```diff
@@ -2684,6 +2733,12 @@ LedgerTxnRoot::Impl::tableFromLedgerEntryType(LedgerEntryType let)
+#ifdef ENABLE_NEXT_PROTOCOL_VERSION_UNSAFE_FOR_PRODUCTION
+    case CONTRACT_DATA:
+        return "contractdata";
+    case CONFIG:
```
`CONFIG_ENTRY` isn't great since we'd wind up with `ConfigEntryEntry`. In @sisuresh's earlier iterations of this he had it named with a mix of "config" and "configsetting". If you like I can change it to `CONFIG_SETTING` and therefore `ConfigSettingEntry` etc.

graydon replied:
Changed to CONFIG_SETTING

graydon commented:
I believe I've now addressed all review comments and have updated the graydon/rs-stellar-xdr/contract-data and stellar/rs-stellar-contract-env/contract-data branches that depend on this (and that this depends on cyclically). 

@jonjove @jayz22 @leighmcculloch @MonsieurNicolas could you please mark comments as resolved that have been resolved?

(and if there's nothing else really critical, can someone r+?)

MonsieurNicolas reviewed (commented):
resolved most comments, I only have one left

jonjove reviewed (commented):
Three very minor comments then let's get this merged.

graydon commented:
Pushed update containing all review comments from @jonjove and @MonsieurNicolas today.

jonjove commented on src/test/TestUtils.cpp:
```diff
@@ -114,6 +115,10 @@ computeMultiplier(LedgerEntry const& le)
         return 1;
+#ifdef ENABLE_NEXT_PROTOCOL_VERSION_UNSAFE_FOR_PRODUCTION
+    case CONTRACT_DATA:
+        return 1;
```
@graydon think you missed this one.

graydon replied:
Oops, fixed.

MonsieurNicolas commented:
r+ f013c999149fd3b9f4176382edcc1f0d6ab4054a
